traits/ctraits.c

```diff
@@ -1816,8 +1816,13 @@ default_value_for(trait_object *trait, has_traits_object *obj, PyObject *name)
             Py_DECREF(tuple);
             if ((result != NULL) && (trait->validate != NULL)) {
                 value = trait->validate(trait, obj, name, result);
-                Py_DECREF(result);
-                return value;
+                if (trait->flags & TRAIT_SETATTR_ORIGINAL_VALUE) {
+                    return result;
```

Comment:
In this branch, we need a check for an error return from the `trait->validate` call: if `value == NULL` at this point, we want to return `NULL` here to propagate the validation error (which will usually be `TraitError`, but could be anything).

(The second branch sneakily avoids the error check: if `value == NULL`, since it's returning `value`, it already returns `NULL` and there's nothing extra to do to propagate the error.)

Reply:
And for extra fun, if `value == NULL`, we also need to `Py_DECREF(result)` before returning our own `NULL`, else we'll have a reference leak.

Reply:
Thanks! I was wondering why the error is behaving so weirdly